codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/generators/TelemetryGenerator.java

```diff
@@ -0,0 +1,91 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.ruby.codegen.generators;
+
+import java.util.Comparator;
+import java.util.Set;
+import java.util.logging.Logger;
+import software.amazon.smithy.codegen.core.directed.ContextualDirective;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.ruby.codegen.GenerationContext;
+import software.amazon.smithy.ruby.codegen.RubyCodeWriter;
+import software.amazon.smithy.ruby.codegen.RubyFormatter;
+import software.amazon.smithy.ruby.codegen.RubySettings;
+import software.amazon.smithy.ruby.codegen.util.Streaming;
+import software.amazon.smithy.utils.StringUtils;
+
+public class TelemetryGenerator extends RubyGeneratorBase {
+    private static final Logger LOGGER =
+            Logger.getLogger(ParamsGenerator.class.getName());
+
+    private final Set<OperationShape> operations;
+
+    public TelemetryGenerator(ContextualDirective<GenerationContext, RubySettings> directive) {
+        super(directive);
+        this.operations = directive.operations();
+    }
+
+    @Override
+    protected String getModule() {
+        return "Telemetry";
+    }
+
+    public void render() {
+        write(writer -> {
+            writer
+                    .preamble()
+                    .includeRequires()
+                    .addModule(settings.getModule())
+                    .apiPrivate()
+                    .addModule("Telemetry")
+                    .call(() -> renderOperations(writer))
+                    .write("")
+                    .closeAllModules();
+        });
+        LOGGER.fine("Wrote telemetry to " + rbFile());
+    }
+
+    private void renderOperations(RubyCodeWriter writer) {
+        operations.stream()
+                .filter((o) -> !Streaming.isEventStreaming(model, o))
+                .sorted(Comparator.comparing((o) -> o.getId().getName()))
+                .forEach(o -> renderOperation(writer, o));
+    }
+
+    private void renderOperation(RubyCodeWriter writer, OperationShape operation) {
+        String classOperationName = symbolProvider.toSymbol(operation).getName();
+        String operationName = RubyFormatter.toSnakeCase(classOperationName);
+        String sdkId = settings.getSdkId();
+
+        writer
+                .write("")
+                .openBlock("class $L", classOperationName)
+                .openBlock("def self.span(context, &block)")
+                .openBlock("attributes = {")
+                .write("'rpc.service' => '$L',", sdkId)
+                .write("'rpc.method' => '$L',", classOperationName)
+                .write("'code.function' => '$L',", operationName)
+                .write("'code.namespace' => '$L'", nameSpace())
+                .closeBlock("}")
+                .openBlock("context.tracer.in_span('$L.$L', attributes: attributes, "
+                        + "kind: Hearth::Telemetry::SpanKind::CLIENT) do",
```

Comment:
Nit: I'd recommend introducing a constant in the Hearth java class for SpanKind.